File: src/ts/magma/app/PathSources.ts
import { Sources } from "../../magma/app/Sources";
import { Path } from "../../magma/api/io/Path";
import { Source } from "../../magma/app/io/Source";
import { Iterable } from "../../magma/api/collect/list/Iterable";
import { IOError } from "../../magma/api/io/IOError";
import { Result } from "../../magma/api/result/Result";
import { PathSource } from "../../magma/app/io/PathSource";
import { ListCollector } from "../../magma/api/collect/list/ListCollector";
export class PathSources implements Sources {
	sourceDirectory: Path;
	constructor (sourceDirectory: Path) {
		this.sourceDirectory = sourceDirectory;
	}
	listSources(): Result<Iterable<Source>, IOError> {
		return this.sourceDirectory().walk().mapValue((children: Iterable<Path>) => {
			return this.retainSources(children)/*unknown*/;
		})/*unknown*/;
	}
	retainSources(children: Iterable<Path>): Iterable<Source> {
		return children.iter().filter((source: Path) => {
			return source.endsWith(".java")/*unknown*/;
		}). < Source > map((child: Path) => {
			return new PathSource(this.sourceDirectory, child)/*unknown*/;
		}).collect(new ListCollector<Source>())/*unknown*/;
	}
}

File: src/java/magma/app/io/source/PathSources.java
package magma.app.io.source;

import magma.api.collect.set.SetCollector;
import magma.api.collect.set.SetLike;
import magma.api.collect.stream.StreamLike;
import magma.api.io.IOError;
import magma.api.io.path.PathLike;
import magma.api.result.Result;

public record PathSources(PathLike root) implements Sources {
    private static SetLike<Source> filter(final StreamLike<PathLike> files) {
        return files.filter(PathLike::isRegularFile)
                .filter(path -> path.asString()
                        .endsWith(".java"))
                .map(PathSource::new)
                .collect(new SetCollector<>());
    }

    @Override
    public Result<SetLike<Source>, IOError> collect() {
        return this.root.walk()
                .mapValue(PathSources::filter);
    }
}